### ISSUE TITLE
SIGPIPE during upgrade to SSL connection

### ISSUE THREAD

**Issue by bdkjones** [warmcat/libwebsockets#447]
I have a Mac app that runs a websocket server using libwebsockets. Here's how I'm creating the server:

```
- (void) startListeningOnPort:(int)port withProtocolName:(NSString *)protocolName andCompletionBlock:(void (^)(NSError *))completionBlock
{
    if (self.isRunning) {
        NSLog(@"Attempt was made to start LPWebSocketServer, but the server is already running. Aborted.");
        return;
    }
  
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0),
    ^{
        lws_set_log_level(LLL_INFO|LLL_DEBUG|LLL_NOTICE|LLL_ERR|LLL_WARN|LLL_HEADER|LLL_EXT|LLL_PARSER|LLL_LATENCY|LLL_COUNT|LLL_CLIENT, logger);
        
        struct lws_protocols protocols[] =
        {
            // first protocol must always be HTTP handler
            {
                "http-only",
                callback_http,
                0,                              // Per-session data size
                0                               // Max frame size
            },
            {
                [protocolName cStringUsingEncoding:NSASCIIStringEncoding],
                callback_websockets,
                sizeof(NSUInteger),            // the session is identified by an id
                0,
                0
            },
            {NULL, NULL, 0, 0}   // End of list
        };
        
        
        int options = LWS_SERVER_OPTION_PEER_CERT_NOT_REQUIRED|LWS_SERVER_OPTION_VALIDATE_UTF8|LWS_SERVER_OPTION_SKIP_SERVER_CANONICAL_NAME;
        
        struct lws_context_creation_info creationInfo = {0};    // Intialize everything in the struct to 0.
        creationInfo.port = 5758;
        creationInfo.iface = NULL;
        creationInfo.protocols = protocols;
        creationInfo.extensions = lws_get_internal_extensions();
        creationInfo.token_limits = NULL;
        creationInfo.ssl_cert_filepath = (_sslCertificatePath) ? [_sslCertificatePath cStringUsingEncoding:NSUTF8StringEncoding] : NULL;
        creationInfo.ssl_private_key_filepath = (_sslKeyPath) ? [_sslKeyPath cStringUsingEncoding:NSUTF8StringEncoding] : NULL;
        creationInfo.ssl_private_key_password = NULL;
        creationInfo.ssl_ca_filepath = NULL;
        creationInfo.ssl_cipher_list = NULL;
        creationInfo.http_proxy_address = NULL;
        creationInfo.http_proxy_port = 0;
        creationInfo.gid = -1;
        creationInfo.uid = -1;
        creationInfo.options = options;
        creationInfo.user = (__bridge void *)(self);           // Pointer to objective-c class/object that holds the socket server.
        creationInfo.ka_time = 0;
        creationInfo.ka_probes = 0;
        creationInfo.ka_interval = 0;
        
        self.context = lws_create_context(&creationInfo);
        
        NSError *error = nil;
        if (self.context == NULL) {
            error = [NSError errorWithDomain:errorDomain code:0 userInfo:@{NSLocalizedDescriptionKey: NSLocalizedString(@"Couldn't create the libwebsockets context.", @"")}];
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
            completionBlock(error);
        });
        
        if (!error)
        {
            self.isRunning = YES;
            
            // For now infinite loop which proceses events and wait for n ms.
            while (!self.stopServer)
            {
                @autoreleasepool
                {
                    lws_service(self.context, 0);
                    
                    if (self.messagesCount > 0) {
                        lws_callback_on_writable_all_protocol(self.context, &protocols[1]);
                    }
                }
                usleep(pollingInterval);
            }
            
            // Once we get here (past the loop), we've been told to stop running the server. Cleanup.
            self.isRunning = NO;
            self.stopServer = NO;
            lws_context_destroy(self.context);
            self.context = NULL;
            [self.messageQueues removeAllObjects];
            self.nextAvailableSessionID = 1;
            self.messagesCount = 0;
            
            dispatch_async(dispatch_get_main_queue(), ^{
                self.serverStoppedCompletionBlock();
                self.serverStoppedCompletionBlock = nil;
            });
        }
    });
}
```

My HTTP callback function is very simple. Right now, it just logs the reason for the callback:

```
static int callback_http(struct lws *wsi,
                         enum lws_callback_reasons reason,
                         void *user,
                         void *in,
                         size_t len)
{
    switch (reason)
    {
        case LWS_CALLBACK_HTTP_WRITEABLE:
        {
            // We have nothing to write over HTTP.
            NSLog(@"HTTP - CALLBACK_HTTP_WRITEABLE");
            return 0;
            break;
        }
        case LWS_CALLBACK_GET_THREAD_ID:
        {
            NSLog(@"HTTP - CALLBACK_GET_THREAD_ID");
            uint64_t tid;
            pthread_threadid_np(NULL, &tid);
            return (int)tid;
            break;
        }
        case LWS_CALLBACK_HTTP:
        {
            NSLog(@"HTTP - CALLBACK_HTTP");
            
            char *requested_uri = (char *) in;
            printf("callback HTTP: requested URI: %s\n", requested_uri);
            
            if (strcmp(requested_uri, "/") == 0) {
                void *universal_response = "Hello, World!";
                lws_write(wsi, universal_response, strlen(universal_response), LWS_WRITE_HTTP);
            }
            return 0;
            break;
        }
        case LWS_CALLBACK_CLOSED_HTTP:
        {
            NSLog(@"HTTP - CALLBACK_CLOSED_HTTP");
            return 0;
            break;
        }
        case LWS_CALLBACK_CHANGE_MODE_POLL_FD:
        {
            NSLog(@"HTTP - CALLBACK_CHANGE_MODE_POLL_FD");
            return 0;
            break;
        }
        case LWS_CALLBACK_UNLOCK_POLL:
        {
            NSLog(@"HTTP - CALLBACK_UNLOCK_POLL");
            return 0;
            break;
        }
        case LWS_CALLBACK_LOCK_POLL:
        {
            NSLog(@"HTTP - CALLBACK_LOCK_POLL");
            return 0;
            break;
        }
        case LWS_CALLBACK_ADD_POLL_FD:
        {
            NSLog(@"HTTP - CALLBACK_ADD_POLL_FD");
            return 0;
            break;
        }
        case LWS_CALLBACK_DEL_POLL_FD:
        {
            NSLog(@"HTTP - CALLBACK_DEL_POLL_FD");
            return 0;
            break;
        }
        case LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED:
        {
            NSLog(@"HTTP - CALLBACK_SERVER_NEW_CLIENT_INSANTITATED");
            return 0;
            break;
        }
        case LWS_CALLBACK_WSI_CREATE:
        {
            NSLog(@"HTTP - CALLBACK_WSI_CREATE");
            return 0;
            break;
        }
        case LWS_CALLBACK_FILTER_NETWORK_CONNECTION:
        {
            NSLog(@"HTTP - CALLBACK_FILTER_NETWORK_CONNECTION");
            return 0;
            break;
        }
        case LWS_CALLBACK_OPENSSL_LOAD_EXTRA_SERVER_VERIFY_CERTS:
        {
            NSLog(@"HTTP - CALLBACK_OPENSSL_LOAD_EXTRA_SERVER_VERIFY_CERTS");
            return 0;
            break;
        }
        case LWS_CALLBACK_PROTOCOL_INIT:
        {
            NSLog(@"HTTP - CALLBACK_PROTOCOL_INIT");
            return 0;
            break;
        }
            
        default:
        {
            NSLog(@"callback_http for reason: %i", reason);
            return 0;
            break;
        }
    }
}
```

# SSL Info

My app runs an HTTPS server on port 5757. That server is secured using the same certificate that I'm using in libwebsockets, above. When the user loads my server address in the browser (https://localhost:5757), they see the "certificate not trusted" warning and must accept the certificate before proceeding. This means that by the time the browser calls the `wss://` link to set up the socket, the user has already manually trusted the certificate, so the call should succeed. 

However, what I'm seeing is that the call to the `wss://` link triggers a bunch of callbacks to my HTTP handler (above) and then I get a SIGPIPE and the app crashes. I believe this is because the client (the browser) is severing the handshake because it doesn't like the self-signed cert. That's fine. But shouldn't libwebsockets be able to handle that gracefully instead of crashing?

# Output

When the websocket context is first created, before any client attempts to connect to it, here's the logged output I get:

```
Initial logging level 1023
Libwebsockets version: 1.7.3 bdkjones@VDK.local-
IPV6 not compiled in
libev support not compiled in
 Threads: 1 each 10240 fds
 mem: platform fd map: 81920 bytes
 mem: per-conn:          400 bytes + protocol rx buf
 Compiled with OpenSSL support
 Using SSL mode
 SSL ECDH curve 'prime256v1'
2016-03-01 10:49:12.908 CodeKit[18612:4108710] HTTP - CALLBACK_OPENSSL_LOAD_EXTRA_SERVER_VERIFY_CERTS
2016-03-01 10:49:12.909 CodeKit[18612:4108710] HTTP - CALLBACK_LOCK_POLL
2016-03-01 10:49:12.909 CodeKit[18612:4108710] HTTP - CALLBACK_ADD_POLL_FD
2016-03-01 10:49:12.909 CodeKit[18612:4108710] HTTP - CALLBACK_UNLOCK_POLL
 Listening on port 5758
2016-03-01 10:49:12.909 CodeKit[18612:4108710] HTTP - CALLBACK_PROTOCOL_INIT
2016-03-01 10:49:12.909 CodeKit[18612:4108710] -----callback_websockets called with reason: 27.
2016-03-01 10:49:12.909 CodeKit[18612:4108710] HTTP - CALLBACK_GET_THREAD_ID
```

And here's the output that occurs when I visit `https://localhost:5757` in my browser, which triggers an attempt to connect to `wss://localhost:5758`:

```
2016-03-01 11:07:33.649 CodeKit[18612:4108710] HTTP - CALLBACK_FILTER_NETWORK_CONNECTION
2016-03-01 11:07:33.649 CodeKit[18612:4108710] HTTP - CALLBACK_WSI_CREATE
2016-03-01 11:07:33.649 CodeKit[18612:4108710] HTTP - CALLBACK_SERVER_NEW_CLIENT_INSANTITATED
2016-03-01 11:07:33.649 CodeKit[18612:4108710] HTTP - CALLBACK_LOCK_POLL
2016-03-01 11:07:33.650 CodeKit[18612:4108710] HTTP - CALLBACK_ADD_POLL_FD
2016-03-01 11:07:33.650 CodeKit[18612:4108710] HTTP - CALLBACK_UNLOCK_POLL
2016-03-01 11:07:33.650 CodeKit[18612:4108710] HTTP - CALLBACK_LOCK_POLL
2016-03-01 11:07:33.650 CodeKit[18612:4108710] HTTP - CALLBACK_CHANGE_MODE_POLL_FD
2016-03-01 11:07:33.650 CodeKit[18612:4108710] HTTP - CALLBACK_UNLOCK_POLL
2016-03-01 11:07:33.652 CodeKit[18612:4108710] HTTP - CALLBACK_LOCK_POLL
2016-03-01 11:07:33.652 CodeKit[18612:4108710] HTTP - CALLBACK_CHANGE_MODE_POLL_FD
2016-03-01 11:07:33.652 CodeKit[18612:4108710] HTTP - CALLBACK_UNLOCK_POLL
2016-03-01 11:07:33.674 CodeKit[18612:4108710] HTTP - CALLBACK_LOCK_POLL
2016-03-01 11:07:33.674 CodeKit[18612:4108710] HTTP - CALLBACK_CHANGE_MODE_POLL_FD
2016-03-01 11:07:33.674 CodeKit[18612:4108710] HTTP - CALLBACK_UNLOCK_POLL
2016-03-01 11:07:33.696 CodeKit[18612:4108710] HTTP - CALLBACK_FILTER_NETWORK_CONNECTION
2016-03-01 11:07:33.696 CodeKit[18612:4108710] HTTP - CALLBACK_WSI_CREATE
2016-03-01 11:07:33.696 CodeKit[18612:4108710] HTTP - CALLBACK_SERVER_NEW_CLIENT_INSANTITATED
2016-03-01 11:07:33.696 CodeKit[18612:4108710] HTTP - CALLBACK_LOCK_POLL
2016-03-01 11:07:33.696 CodeKit[18612:4108710] HTTP - CALLBACK_ADD_POLL_FD
2016-03-01 11:07:33.696 CodeKit[18612:4108710] HTTP - CALLBACK_UNLOCK_POLL
2016-03-01 11:07:33.696 CodeKit[18612:4108710] HTTP - CALLBACK_LOCK_POLL
2016-03-01 11:07:33.696 CodeKit[18612:4108710] HTTP - CALLBACK_CHANGE_MODE_POLL_FD
2016-03-01 11:07:33.696 CodeKit[18612:4108710] HTTP - CALLBACK_UNLOCK_POLL
2016-03-01 11:07:33.698 CodeKit[18612:4108710] HTTP - CALLBACK_LOCK_POLL
2016-03-01 11:07:33.698 CodeKit[18612:4108710] HTTP - CALLBACK_CHANGE_MODE_POLL_FD
2016-03-01 11:07:33.698 CodeKit[18612:4108710] HTTP - CALLBACK_UNLOCK_POLL
2016-03-01 11:07:33.698 CodeKit[18612:4108710] HTTP - CALLBACK_CHANGE_MODE_POLL_FD
2016-03-01 11:07:33.698 CodeKit[18612:4108710] HTTP - CALLBACK_CLOSED_HTTP
2016-03-01 11:07:33.699 CodeKit[18612:4108710] HTTP - CALLBACK_LOCK_POLL
2016-03-01 11:07:33.699 CodeKit[18612:4108710] HTTP - CALLBACK_CHANGE_MODE_POLL_FD
2016-03-01 11:07:33.699 CodeKit[18612:4108710] HTTP - CALLBACK_UNLOCK_POLL
2016-03-01 11:07:33.720 CodeKit[18612:4108710] HTTP - CALLBACK_LOCK_POLL
2016-03-01 11:07:33.720 CodeKit[18612:4108710] HTTP - CALLBACK_DEL_POLL_FD
2016-03-01 11:07:33.720 CodeKit[18612:4108710] HTTP - CALLBACK_CHANGE_MODE_POLL_FD
2016-03-01 11:07:33.720 CodeKit[18612:4108710] HTTP - CALLBACK_UNLOCK_POLL
(App terminates with SIGPIPE)
```

In Chrome, the Console shows this error message:

```
WebSocket connection to 'wss://localhost:5758/' failed: WebSocket opening handshake was canceled
```

That's what's leading me to believe Chrome is rejecting the connection once the process gets underway and libwebsockets is then attempting to write to a closed pipe, which generates the SIGPIPE. 

Have I done something wrong? I certainly wouldn't rule that out. Note that if I don't use SSL, everything works just fine.

**Comment by bdkjones** [comment]
An interesting development. I added the handler `signal(SIGPIPE, SIG_IGN);` to skip the SIGPIPE, then allowed execution to continue past the point where the SIGPIPE arrives (Xcode's debugger always breaks on SIGPIPE, even when the ignore handler is in place.) 

The websocket connection actually works! I'm able to send/receive data across the `wss://` socket.

Obviously, though, the SIGPIPE that happens during the initial setup is something I'd like to solve rather than simply throw in a hack to ignore it. Any ideas what's causing this, or how I can go about diagnosing it? I'm not familiar with the libwebsockets implementation, obviously.

**Comment by bdkjones** [comment]
I've updated the title because the SIGPIPE is happening during the handshake/upgrade process, but that process is not, apparently, failing.

**Comment by lws-team** [comment]
Is the same situation happening with the test server + client that are built along with the library by default?  You can run them with --ssl to use a selfsigned cert generated at cmake-time.

**Comment by bdkjones** [comment]
I don't know because I can't get your full suite to build on OS X using LLVM through Xcode. I get a linker error on the .dylib target. To get the static library (*.a) to build, I had to remove the dynamic library target and all the test products (since they depend on the dylib). 

Here's the linker error I'm seeing. Note that I select Xcode as my build tool in CMake and the only option I toggled in CMake was to turn on IPV6 support. Otherwise, everything is default. 

![screen shot 2016-03-01 at 18 34 00](https://cloud.githubusercontent.com/assets/926566/13449047/90972488-dfdc-11e5-96aa-b4ae99fac855.png)

**Comment by lws-team** [comment]
How old is your openssl?  This api should be provided by your openssl library.

If your openssl is really very old, and you can't update it, you can try replace lws use of that api with

ERR_remove_state(0);

...which is the deprecated, old version of the same api.  See #186

**Comment by bdkjones** [comment]
Hmm. It looks like CMake is using the executable at `/usr/bin/openssl` which is the one that ships with OS X. It's version 0.9.8zg 14 July 2015.

The include path I gave CMake for OpenSSL is my own installation of OpenSSL 1.0.2f 28 Jan 2016. Perhaps that is the trouble. I'll reconfigure CMake to point to this installation's executable and see if that makes a difference.

**Comment by lws-team** [comment]
Yeah.  But it's not the 'executable', it's the related openssl library path you need to override.

You can find info on how to do it in the README.build

https://github.com/warmcat/libwebsockets/blob/master/README.build.md

**Comment by bdkjones** [comment]
Got the test suite to build. The test-server does not exhibit a SIGPIPE during the SSL upgrade, as far as I can tell. I attached LLDB to it, to be sure, and it appears to run normally.

Given that, how would you suggest I track down the exact write operation that's triggering the SIGPIPE in the situation above?

**Comment by lws-team** [comment]
I think the write() is OK because there may not be a way to discover this situation without writing to it.

Actually lws captures and ignores SIGPIPE early in the lws_create_context() code.  After that you shouldn't get any SIGPIPE from the whole process, he'll discover it later as an error state on poll().  So there is something different in your code to the test server with regards to this.

**Comment by bdkjones** [comment]
Well, that cleared it up. When I originally hit "configure" in CMake, it warned me that it could not find the OpenSSL include path. I manually set that to `/usr/local/ssl/include` but I did NOT update the paths to `libcrypto`, `libssl` and the OpenSSL executable --- those were still set to use the system default installation of OpenSSL.

I changed the latter paths to all point to the 1.0.2f installation of OpenSSL, then regenerated with CMake. All targets then built properly in Xcode. I rebuilt my app using the new `libwebsockets.a` and now the SIGPIPE does not occur during the upgrade process. So the issue was definitely the older version of OpenSSL (or the mismatched include path from one version and libraries from an older version).

Sorry for the trouble!
